src/wp-includes/compat.php

```diff
@@ -478,6 +478,33 @@ function str_ends_with( $haystack, $needle ) {
 	}
 }
 
+if ( ! function_exists( 'array_is_list' ) ) {
+	/**
+	 * Polyfill for `array_is_list()` function added in PHP 8.1.
+	 *
+	 * Determines if the given array is a list.
+	 *
+	 * An array is considered a list if its keys consist
+	 * of consecutive numbers from 0 to count( $array ) - 1.
+	 *
+	 * This function returns `true` on empty arrays.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array $arr The array.
+	 * @return bool True if array is a list, false otherwise.
+	 */
+	function array_is_list( array $arr ) {
+		$i = 0;
+		foreach ( $arr as $k => $v ) {
```

Comment:
1. This uses `foreach ( $arr...` rather than `array_keys( $arr )` or `array_keys( $arr ) === range( count( $arr ) - 1 )` to avoid creating new arrays unnecessarily. This means only one pass over the array, and is in line with the example given in the [RFC's Proposal section](https://wiki.php.net/rfc/is_list).
2. An early return for an empty array is not included because an empty array will skip on the `foreach()` and return `true`.

Reply:
Looking over this more, I wonder if this is actually enough to match [the docs](https://www.php.net/manual/en/function.array-is-list.php):

```php
function array_is_list( array $arr ) {
    return array() === $arr || array_values( $arr ) === $arr;
}
```

This implementation is [used by Symfony](https://github.com/symfony/polyfill/blob/main/src/Php81/Php81.php#L21-L36), but they also have more for when these conditions aren't matched. However, I haven't identified a case where the conditions aren't matched.

Current implementation: https://3v4l.org/tgUEX/perf
Alternative implementation: https://3v4l.org/MLs3Z/perf